components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,394 @@
+        let mut buf = Vec::new();
+        f.read_to_end(&mut buf)?;
+        let mut remained = buf.as_slice();
+
+        // parse the file dictionary
+        let (header, content) = Header::parse(remained)?;
+        remained.consume(Header::SIZE + content.len());
+        let mut file_dict = FileDictionary::default();
+        match header.version() {
+            Version::V1 => {
+                let mut encrypted_content = EncryptedContent::default();
+                encrypted_content.merge_from_bytes(content)?;
+                let content = PlaintextBackend::default().decrypt(&encrypted_content)?;
+                file_dict.merge_from_bytes(&content)?;
+            }
+            Version::V2 => {
+                file_dict.merge_from_bytes(content)?;
+                // parse the remained records
+                while !remained.is_empty() {
+                    // If the parsing get errors, manual intervention is required to recover.
+                    let (used_size, file_name, mode) = Self::parse_next_record(remained)?;
+                    remained.consume(used_size);
+                    match mode {
+                        LogRecord::INSERT(info) => {
+                            file_dict.files.insert(file_name, info);
+                        }
+                        LogRecord::REMOVE => {
+                            let original = file_dict.files.remove(&file_name);
+                            if original.is_none() {
+                                return Err(box_err!(
+                                    "Try to recovery from log file but remove a null entry, file name: {}",
+                                    file_name
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(file_dict)
+    }
+
+    /// Append an insert operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn insert(&mut self, name: &str, info: FileInfo) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::INSERT(info))?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
+                "file corrupted! write operation must be called before append."
+            ))
+        }
+    }
+
+    /// Append a remove operation to the log file.
+    ///
+    /// Warning: `self.write(file_dict)` must be called before.
+    pub fn remove(&mut self, name: &str) -> Result<()> {
+        if let Some(file) = self.append_file.as_mut() {
+            let bytes = Self::convert_record_to_bytes(name, LogRecord::REMOVE)?;
+            file.write_all(&bytes)?;
+            file.sync_all()?;
+            self.check_compact()?;
+            Ok(())
+        } else {
+            Err(box_err!(
```

Comment:
Seems all operations need the append file. So I prefer `assert!` to make it clear.